### PR TITLE
Enhancement: Allow org block live sync

### PR DESCRIPTION
**Pull request by devcarbon-com**
Draft pull request.  Preliminary manual testing has been successful. 

Not sure when org started support for it, but it works with 9.7-pre, and hopefully 9.6 also.

### CODE REVIEW

**Comment by devcarbon-com** [comment]
This seems to work okay, but I feel like it's a bit hacky and solving the problem at the wrong location. 

I think it would be better to add 2 more markers to the text properties in the case of org blocks.

Right now, we have `org-transclusion-beg-mkr` and `org-transclusion-end-mkr` that covers the whole range of the content added when the transclusion is expanded.  But these do not distinguish between exact content, and modified content (IE. when we add `#begin_src lang` and `#end_src`). 

One way to do this without having to change a lot of other code would be to add text properties specifying the orginal content range when we modify the `src-content` in  `org-transclusion-content-src-lines` and then in `org-transclusion-content-insert` to extract them from `content` and add markers.


I have a proof-of-concept of this locally on my machine, but I'm not aware of how to best proceed when it comes to pull requests that become inter-dependent. 

@nobiot Should I just add it as it's own request, and then merge this (#176) and #157 after it's accepted?

**Comment by devcarbon-com** [comment]
@nobiot Sure thing!

I'm happy to add a code example if you would like. Sometimes it's easier to read code than English :P.

**Comment by nobiot** [comment]
Sorry, I think I do not seem to have understood the intent of this PR.
See below... I try starting live-sync with the test files below; no change of behaviour.

It will be great if you could provide some example files.

image


``` org
#test-live-sync.org
#+transclude: [[file:test-src.org::code]]
```

``` org
#test-src.org
#+name: code
#+begin_src emacs-lisp transclude file:simple.el :lines 1-10 -n  -r
         (save-excursion                 ;; (ref:sc)
            (goto-char (point-min))
            (message "executed."))     ;; (ref:jump)
#+end_src
```

**Comment by devcarbon-com** [comment]
Oh I had not tried transcluding code from another org file.  I'd need to do some digging to figure out how to do that. 

What I have been doing is like this:

Narrative.org
```elisp
#+transclude: [[file:code.el]]  :src elisp
```
code.el
```elisp
(defun hello ()
  (message "hello"))
```

**Comment by nobiot** [comment]
Oh, now I got it!
All what I said in [this discussion thread](https://github.com/nobiot/org-transclusion/discussions/114#discussioncomment-5201326) is about transcluding a source code block from another Org file -- i.e. the source being another Org file. 

The use case for this PR is transcluding a text file, it is not applicable -- I'm sure you figured that out.

I have tested the code lightly with my test python files (I should perhaps put into the repo as a common test file set). Your code change works extremely well as far as I could see -- I could even use `org-edit-src` when live-sync is on. Whoa. 

The reason why I didn't let it work by this:

```
(when (string= "src" type)
  (user-error "No live sync for src-code block"))
```

is that I could not work out a reliable way to deal with the indentation -- the length of the source and transclusion must exactly the same -- I think you have figured out a way.

**Comment by nobiot** [comment]
There is one suggestion, if I may. I will write it directly in the review of the source code as well.  Could you look into moving the change you made in function `org-transclusion-live-sync-buffers-others-default` to function `org-transclusion-live-sync-buffers-src-lines` in file `org-transclusion-src-lines.el` by any chance? This way:

1. We can contain the change for "src" type transclusion to a function
2. Probably you can remove `(if (org-in-src-block-p)` clause (because "src" type always results in the source text file be in a src block

What do you think?

**Review by nobiot — CHANGES_REQUESTED** [review]
If you could consider the change I suggested, that will be great. I have a feeling it should be merged as soon as you let me know the suggested change is not feasible, or you have changed it according my suggestion -- I am really hoping that the suggested change is feasible...

File: org-transclusion.el [hunk 1]
```diff
@@ -1658,16 +1672,16 @@ ensure the settings revert to the user's setting prior to
...
-    ;; if not push to elm the list
-    ;; wrap-prefix, etc.
-    (dolist (sym org-transclusion-yank-excluded-properties)
-      (if (memq sym yank-excluded-properties)
-        (push sym org-transclusion-yank-remember-user-excluded-props)
-        ;; Avoid duplicate
-        (push sym excluded-props)))
-    (setq yank-excluded-properties
-          (append yank-excluded-properties excluded-props)))))
+      ;; Return t if 'wrap-prefix is already in `yank-excluded-properties'
```
[the remark this answers is not in the document]
Comment (devcarbon-com) [reply org-transclusion.el:1675]: Oh oops, yes that is exactly all it is.  I must have bumped it somehow.   I probably should disable aggressive indent while working on pull requests. Or maybe figure out a way to make it match the style already used in the file.

File: org-transclusion.el [hunk 2]
```diff
@@ -1637,9 +1640,20 @@ This function is for non-Org text files."
                        (overlay-start tc-pair)
                        (overlay-end tc-pair)
                        (overlay-buffer tc-pair)))
-              (tc-ov (text-clone-make-overlay
-                      (get-text-property (point) 'org-transclusion-beg-mkr)
-                      (get-text-property (point) 'org-transclusion-end-mkr))))
+              (tc-ov-beg-mkr (get-text-property (point) 'org-transclusion-beg-mkr))
+              (tc-ov-end-mkr (get-text-property (point) 'org-transclusion-end-mkr))
+              (tc-ov (if (org-in-src-block-p)
```
[the remark this answers is not in the document]
Comment (devcarbon-com) [reply org-transclusion.el:1645]: I'll give it a shot!  :)

File: org-transclusion.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (nobiot) [reply org-transclusion.el:1675]: No worries :)

**Comment by devcarbon-com** [comment]
@nobiot Oh, I thought I just had to check the box, so I'm glad you brought it up!

Got it started now.

**Comment by nobiot** [comment]
Thank you!

**Comment by nobiot** [comment]
> That's definitely doable, but I'm not sure how in elisp do that without it being rather verbose, so I wrote a macro that I personally use a lot, but I'm not sure it would be a good idea to use in production code. (I'm a elisp newbie).

I have an idea; a simple one without a macro. Let me get back to you as a suggestion. Please give me until the end of the week (I am hoping to have some time tomorrow but if it fails I need a bit more time).

**Comment by nobiot** [comment]
@devcarbon-com I have no idea how I can send a commit to your PR so I opened a new one as continuation #178. Your original commits are there, too. Could you take a look? 

To be honest, I don't understand the "dot" syntax you show here: `(+ .a .b .c)` and how the macro works (I'm not familiar with `let-alist`. 

My suggestion is simple and probably crude compared to your macro-based way: it is just to run the default function to create the src-ov and tc-ov, and o move tc-ov's location based on your offset logic... Kinda avoids "repeating yourself" issue, I guess. But not sophisticated. Perhaps this is sufficient?

Let me know of your thoughts...  At least, functionally it seems to works in the same way as yours. 

```
(defun do-the-work ()
  (alist-of-let* ((a 1)
                  (b 2)
                  (c (- b a))))

(defun delegate-the-work ()
  (let-alist (do-the-work)
    (+ .a .b .c)))

(delegate-the-work) ;=> 4
```

**Pull request merged**